### ISSUE TITLE
healpixpad_cuda module not available after successfully built earth2grid

### ISSUE THREAD

**Issue by einzigsue** [NVlabs/earth2grid#37]
Is building the module `healpixpad_cuda` part of the standard installation? If not, how should I build it as part of earth2grid? 

With python 3.12 and gcc/gxx 14.1.0, I installed earth2grid by running the command below
```
pip3 install --no-build-isolation git+https://github.com/NVlabs/earth2grid.git@v2025.4.1
```

and the build log says

```
...
Collecting git+https://github.com/NVlabs/earth2grid.git@v2025.4.1
  Cloning https://github.com/NVlabs/earth2grid.git (to revision v2025.4.1) to /tmp/pip-req-build-22mrojon
  Running command git clone --filter=blob:none --quiet https://github.com/NVlabs/earth2grid.git /tmp/pip-req-build-22mrojon
  Running command git checkout -q baa00d28e647fc731dfaa44ee45a86131d50bcf5
  Resolved https://github.com/NVlabs/earth2grid.git to commit baa00d28e647fc731dfaa44ee45a86131d50bcf5
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting einops>=0.7.0 (from earth2grid==2025.4.1)
  Downloading einops-0.8.1-py3-none-any.whl.metadata (13 kB)
Requirement already satisfied: netCDF4>=1.6.5 in /opt/conda/envs/cbottle/lib/python3.12/site-packages (from earth2grid==2025.4.1) (1.7.2)
Requirement already satisfied: numpy>=1.23.3 in /opt/conda/envs/cbottle/lib/python3.12/site-packages (from earth2grid==2025.4.1) (2.2.6)
...
Building wheels for collected packages: earth2grid
  Building wheel for earth2grid (pyproject.toml): started
  Building wheel for earth2grid (pyproject.toml): finished with status 'done'
  Created wheel for earth2grid: filename=earth2grid-2025.4.1-cp312-cp312-linux_x86_64.whl size=119700 sha256=6522fbd399a498e034f9a9c9915bf2e4037523e849179375d935f11a888fa2ef
  Stored in directory: /tmp/pip-ephem-wheel-cache-7bdyczpx/wheels/e5/75/08/1f313805620739f150931bdc1d933fef1f29e71e1acb70a4ee
Successfully built earth2grid
```

but when importing earth2grid it says submodule `healpixpad_cuda` not available. 
```
>>> import earth2grid
/opt/conda/envs/cbottle/lib/python3.12/site-packages/earth2grid/healpix.py:61: UserWarning: healpixpad_cuda module not available, reverting to CPU for all padding routines
  warnings.warn("healpixpad_cuda module not available, reverting to CPU for all padding routines")
```

**Comment by einzigsue** [comment]
Thank you @nbren12 for the reply! I now understand. Because I built the package on a CPU only node, `pip3 install --no-build-isolation git+https://github.com/NVlabs/earth2grid.git@v2025.4.1` didn't pick up the task to build `healpixpad_cuda`.

Is it a reason why the build has to be on a GPU node? Is it the CUDA implementation requires a minimal GPU capability? for example H100 and later ones? Can you point to where the code finds out the GPU availability?  Unfortunately my CI/CD pipeline is set up on a CPU only server.

**Comment by nbren12** [comment]
> The healpixpad extension should automatically compile [if CUDA is present](https://github.com/NVlabs/earth2grid/blob/89a71d2c86ed8ebbdf1f2e814c02d2966276e46f/setup.py#L84).

This link shows where CUDA availability is detected. Basically if it fails to import the needed code from torch. I doubt the GPU capability is an issue since I have run it on a variety of cards. 

You may be able to install CUDA without the drivers, but I haven't done it. The simplest way to make it work is to compile it against the specific CUDA and pytorch versions present in the compute environment. 

Could you say more about what you are trying to accomplish? As I said before, the pure python implementation of the padding should be as fast as the CUDA, which makes this discussion somewhat moot.

**Comment by einzigsue** [comment]
@nbren12 thank you for the prompt reply! It is reassuring to hear that the performance between the two padding implements are not very different.

For the pure sake of understanding the build process:) I double checked my conda environment and I do have `CUDAExtension` from torch but for whatever reason earth2grid doesn't build its healpixpad_cuda module. 

```
>>> import torch
>>> torch.cuda.is_available()
True
>>> from torch.utils.cpp_extension import CUDAExtension
>>> import earth2grid
/opt/conda/envs/cbottle/lib/python3.12/site-packages/earth2grid/healpix.py:61: UserWarning: healpixpad_cuda module not available, reverting to CPU for all padding routines
  warnings.warn("healpixpad_cuda module not available, reverting to CPU for all padding routines")
```
and looking into my build script, I did install `torch` before `earth2grid`
```
conda create -n cbottle --yes gcc_linux-64=14.1.0 gxx_linux-64=14.1.0 python=3.12 wget nvtop gotop && \
conda activate cbottle && \
mamba install -c conda-forge healpy pyshtools shapely geopandas rich gcsfs xarray[complete] matplotlib zarr netCDF4 h5netcdf scipy scikit-learn s3fs zict intake intake-esm jupyterlab jupyterlab-nvdashboard dask distributed dask-labextension jupyterlab-favorites && \
pip3 install torch>=2.5.0 torchvision torchaudio && \
pip3 install scores cartopy pytorch-lightning lightning && \
pip3 install easygems  && \
pip3 install setuptools hatchling && \
pip3 install --no-build-isolation git+https://github.com/NVlabs/earth2grid.git@v2025.4.1 && \
pip3 install git+https://github.com/NVlabs/cBottle.git@main
```

Any thoughts why the build process doesn't pick up installing the `healpixpad_cuda` module?

**Comment by nbren12** [comment]
First, I would recommend our dockerfile: https://github.com/NVlabs/cBottle/blob/main/Dockerfile and build a machine with a GPU. This is based off our [pytorch images](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch) which have everything needed to get started with building earth2grid (or any pytorch CUDA extension).


That said, perhaps we can get your conda env working:

> Any thoughts why the build process doesn't pick up installing the healpixpad_cuda module?

Maybe you need to install the CUDA compilers? You shown there is no ImportError, but the setup.py throws an OSError if the compiler toolchain isn't installed. 
```
try:
    from torch.utils.cpp_extension import CUDAExtension

    ext_modules.append(
        CUDAExtension(
            name='healpixpad_cuda',
            sources=cuda_src_files,
            extra_compile_args={'nvcc': ['-O2']},
        ),
    )
except (ImportError, OSError):
    warnings.warn("Cuda extensions for torch not found, skipping cuda healpix padding module")
```
Here are [some instructions](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#installing-cuda-using-conda) on installing CUDA in a conda env. I don't use conda much myself, so unfortunately cannot help any more here. I find the containers work well.

**Comment by nbren12** [comment]
note to self:

- [ ] add more helpful error message when OSError raised
- [ ] Document the need for a CUDA installation that includes compilers. The pip wheels do not contain these.

**Comment by einzigsue** [comment]
@nbren12 Thanks a lot! 
Now I know how to fix it:) have a nice weekend.
